src/lib/shim/shadow_sem.c

```diff
@@ -0,0 +1,173 @@
+#include "lib/shim/shadow_sem.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <linux/futex.h>
+#include <semaphore.h>
+#include <stdbool.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+typedef struct {
+    // Upper bound of how many wake operations have occurred.
+    uint16_t nwakes;
+    // How many waiting on the semaphore. Used to avoid unnecessary futex-wake operations.
+    uint16_t nwaiters;
+} wakes_t;
+
+typedef struct {
+    union {
+        struct {
+            // We use `wakes` as the futex word, which must be exactly 4 bytes long and
+            // 4-byte-aligned.
+            wakes_t wakes;
+            // Value of the semaphore.
+            uint32_t value;
+        };
+        uint64_t __align;
+    };
+} shadow_sem_impl_t;
+
+static inline shadow_sem_impl_t* _impl(shadow_sem_t* sem) {
+    assert(sizeof(shadow_sem_impl_t) == sizeof(shadow_sem_t));
+    assert(sizeof(wakes_t) == 4);
+    return (shadow_sem_impl_t*)sem;
+}
+
+static int _futex(int* uaddr, int futex_op, int val, const struct timespec* timeout, int* uaddr2,
+                  int val3) {
+    return (int)syscall(SYS_futex, uaddr, futex_op, val, timeout, uaddr2, val3);
+}
+
+int shadow_sem_init(shadow_sem_t* _sem, int pshared, unsigned int value) {
+    shadow_sem_impl_t* sem = _impl(_sem);
+    if (value > SEM_VALUE_MAX) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    *sem = (shadow_sem_impl_t){
+        .wakes =
+            {
+                .nwakes = 0,
+                .nwaiters = 0,
+            },
+        .value = (int)value,
+    };
+
+    return 0;
+}
+
+int shadow_sem_destroy(shadow_sem_t* sem) {
+    // Nothing to do.
+    return 0;
+}
+
+int shadow_sem_post(shadow_sem_t* _sem) {
+    shadow_sem_impl_t* sem = _impl(_sem);
+    shadow_sem_impl_t sem_prev_value;
+    __atomic_load(sem, &sem_prev_value, __ATOMIC_RELAXED);
```

Comment:
> p.s. I'm noticing that you're not using <stdatomic.h> --- is this because we have to support GCC 4.8.5?

Right :/

> it might be worth a lightweight port to C++11 so that we can use the standard library.

Maybe, but then we'd definitely need an opaque wrapper type to expose to the C code, which you were hoping to get rid of :).

I do think using a more modern atomics api is probably worth it though - wdyt?

Reply:
> I do think using a more modern atomics api is probably worth it though - wdyt?

Yep, you are right. The best compromise here is to use C++11's atomics in the standard library and declare an opaque struct for C. I think the change from `shadow_sem.c` to `shadow_sem.cc` shouldn't be too bad... agreed?

Reply:
Since we're dropping centos 7, I used C11 atomics, since I figured that'd be a little less work than changing to C++.

It turned out to be a little awkward with the design I had, because fields of `_Atomic` struct types can't be accessed without atomic operations. So e.g. after doing an atomic load of the whole semaphore, I still needed to do more atomic operations to access the futex sub-structure. Since those were already on a local copy I could use 'relaxed' operations, which I *think* are cheap (?), but it was still awkward.

I ended up rethinking it a bit and I think landed on a better design. I ended up using `memory_order_seq_cst` to ensure a consistent ordering between the operations on `_value` and `_nwaiters` instead of the previous approach of tracking the total number of wakes and jamming that into the futex. It's possible we could use a weaker memory order than seq_cst, but by my measurements we're already getting about the same performance as with the glibc semaphore, so I'm inclined to just leave it using that.